sw/airborne/modules/stereocam/stereocam2state/stereocam2state.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) Kimberly McGuire
+ *
+ * This file is part of paparazzi
+ *
+ */
+/**
+ * @file "modules/stereocam/stereocam2state/stereocam2state.c"
+ * @author Kimberly McGuire
+ * This module sends the data retreived from an external stereocamera modules, to the state filter of the drone. This is done so that the guidance modules can use that information for couadcopter
+ */
+
+#include "modules/stereocam/stereocam2state/stereocam2state.h"
+#include "modules/stereocam/stereocam.h"
+
+#include "subsystems/abi.h"
+#include "state.h"
+
+#include "math/pprz_algebra_int.h"
+#include "math/pprz_algebra_float.h"
+#include "math/pprz_orientation_conversion.h"
+#include "subsystems/gps.h"
+
+#ifndef SENDER_ID
+#define SENDER_ID 1
+#endif
+#ifndef USE_DEROTATION_OPTICFLOW
+#define USE_DEROTATION_OPTICFLOW FALSE
+#endif
+#ifndef STATE_MEASURE_OPTICFLOW
+#define STATE_MEASURE_OPTICFLOW TRUE
+#endif
+
+static float prev_phi;
+static float prev_theta;
+
+
+
+void stereocam_to_state(float dphi, float dtheta);
+
+void stereo_to_state_init(void) {}
+void stereo_to_state_periodic(void)
+{
+  if (stereocam_data.fresh) {
+    stereocam_data.fresh = 0;
+    float phi = stateGetNedToBodyEulers_f()->phi;
+    float theta = stateGetNedToBodyEulers_f()->theta;
+    float dphi =  phi - prev_phi;
+    float dtheta = theta - prev_theta;
+
+    stereocam_to_state(dphi, dtheta);
+
+    prev_theta = theta;
+    prev_phi = phi;
+  }
+}
+void stereo_to_state_start(void) {}
+void stereo_to_state_stop(void) {}
+
+
+void stereocam_to_state(float dphi, float dtheta)
+{
+
+  // Get info from stereocam data
+  float vel_hor = ((float)(stereocam_data.data[8]) - 127) / 100;
+  float vel_ver = ((float)(stereocam_data.data[9]) - 127) / 100;
+  float vel_x = 0;
+  float vel_y = 0;
+
+  // Calculate derotated velocity
+#if USE_DEROTATION_OPTICFLOW
+  float agl_stereo = (float)(stereocam_data.data[4]) / 10;
+
+
+  float diff_flow_hor = dtheta * 128 / 1.04;
+  float diff_flow_ver = dphi * 96 / 0.785;
+
+  float diff_vel_hor = diff_flow_hor * agl_stereo * 12 * 1.04 / 128;
+  float diff_vel_ver = diff_flow_ver * agl_stereo * 12 * 0.785 / 96;
+
+  vel_x = - (vel_ver - diff_vel_ver);
+  vel_y = (vel_hor - diff_vel_hor);
+#endif
+
+  // Derotate velocity and transform from frame to body coordinates
+  vel_x = - (vel_ver);
+  vel_y = (vel_hor);
+
+
+#if STATE_MEASURE_OPTICFLOW
+  // Calculate velocity in body fixed coordinates from opti-track and the state filter
+  struct NedCoor_f coordinates_speed_state;
+  coordinates_speed_state.x = stateGetSpeedNed_f()->x;
+  coordinates_speed_state.y = stateGetSpeedNed_f()->y;
+  coordinates_speed_state.z = stateGetSpeedNed_f()->z;
+
+  struct NedCoor_f opti_state;
+  opti_state.x = (float)(gps.ecef_vel.y) / 100;
```

Comment:
On a sidenote: accessing the GPS struct directly is kind of deprecated and won't be possible any more soon(-ish)...
Should be changed to ABI subscription of the GPS message....